### ISSUE TITLE
Heavy performance impact from basic actions such as scroll, hover, and click

### ISSUE THREAD

**Issue by epiqu1n** [craigmckeachie/handsonreact.com#3]
<h2>Issue</h2>

I have noticed that performing basic actions such as scrolling a single tick (let alone rapid scrolling), clicking anywhere on the page, or hovering over links, is causing significant spikes in CPU usage. I monitored through Windows Task Manager as well as Firefox Task Manager, and both reflected this finding. Please let me know if there is any information in addition to the following that I can provide.

<h2>Data</h2>

What initially tipped me off was that while browsing the page [State | HandsOnReact](https://handsonreact.com/docs/state), I kept hearing my computer's fans spinning up, and realized after some time that scrolling on the page was consistently causing it. After playing around with it a bit, I took a video recording of me performing basic actions on the page with the Windows and Firefox task manager's visible. I also captured a performance profile using Firefox's Profiler to go alongside it:

[HandsOnReact Performance Test Video](https://www.youtube.com/watch?v=vdiMy2u6sbg)
[Firefox Performance Profile](https://share.firefox.dev/3iRKncR)

Shown in the video is first, after beginning the performance recording, me clicking outside the window then back in which causes a small jump in CPU/energy usage. I then begin scrolling slowly in discrete ticks, which causes a jump in CPU usage each time. Then during rapid scrolling, not only does CPU usage increase even more, but there is a period of a few seconds where the page disappears entirely from Firefox Task Manager and CPU usage remains high - despite scrolling having stopped. Lastly, I hover back and forth over the page heading links in the sidebar which causes a similar behavior to the scrolling (and a noticeable delay can be seen in when the links respond to being hovered over).

I performed similar tests on a Wikipedia page and did not notice similar behavior at all.

<h2>System Specs</h2>
I am using Firefox version 98.0.2 (64-bit) on Windows 10 Pro version 21H2. My system specs are:

- Intel i7-8700k @ 4.7 GHz
- G.Skill 32 GB RAM
- EVGA GTX 1080 Ti

**Comment by epiqu1n** [comment]
As I have been continuing to use the site, I have noticed that pages start out much more responsive. However, the longer the page stays open (or perhaps the more it has been interacted with), the less responsive it becomes with any kind of interaction.

**Comment by craigmckeachie** [comment]
Thanks for taking the time to report this and provide detailed information on reproducing the issue.

I have had others report this before and it almost always seems to be on Firefox.  If you have the SetPoint extension please try to disable the setpoint Firefox extension and let me know if the issue still persists.

https://addons.mozilla.org/en-US/firefox/addon/logitechsmoothscrolling/
See this for more details.
https://support.mozilla.org/en-US/questions/1048084

I am using Docusaurus for the site
https://docusaurus.io/
Which is what the React documentation site uses and several other React community projects use, however I'm using a version 2 early alpha version that I need to upgrade but that is pretty involved but I will eventually get to it.
Another site to try to reproduce the behavior on that uses Docusaurus version 2 is:
https://redux.js.org/

**Comment by epiqu1n** [comment]
Of course, I find it interesting looking into stuff like this and am happy to post my findings in case they can be of any use. For the record, I do not use SetPoint and/or Logitech Smooth Scrolling extension; [these are the only extensions](https://i.imgur.com/WEbbQFz.png) that I use.

I just tried playing around with the Redux link you provided and it does not seem to exhibit the same behavior. Interestingly, the idle Energy Impact reported by Firefox seems to be abnormally high for Redux's site at 10 (of some unit), compared to HandsOnReact which has *no* measurable impact while idle (or even Netflix which idles at 3.5 with a paused video open). The total CPU usage for Firefox hovers around 1% for either page when idle.

However, during rapid scrolling, while Redux's page only shows a slight increase - up to 12 Energy Impact and ~11% CPU usage, HandsOnReact spikes and disappears from the Energy Impact list and shows about ~20% CPU usage. In fact, even scrolling relatively slowly causes quite high spikes in HandsOnReact while the change is much lower on Redux.

That said, I definitely understand the pains of upgrading and I wouldn't worry about it *too* much. The site is certainly still usable and a refresh will alleviate slowness for long enough for it to not be very problematic. Not to mention it's been an excellent resource for learning React 😄
